### PR TITLE
BUG: fix handling for `factorial(..., exact=False)` for 0-dim array inputs

### LINKED ISSUE #18753: BUG: factorial return type inconsistent for 0-dim arrays
While I tried to be thorough in #15600 and #15841, I got caught by the fact that `np.array` will ignore an explicit `ndim=0` request and happily give back a 1-dimensional result without complaint:
```
>>> np.array([1], ndmin=0, dtype=np.int64).ndim
1
```

This ends up breaking the assumptions in my `test_factorial{,2,k}_array_corner_cases` tests... 😑

### PR DESCRIPTION
**Pull request by h-vetinari**
Oversight from #15841.

~Start by adding a failing test for #18753~

Fixes #18753

### CODE REVIEW

File: scipy/special/tests/test_basic.py
Comment (h-vetinari) [inline scipy/special/tests/test_basic.py:2157]:
```suggestion
        # assert_equal does not distinguish scalars and 0-dim arrays of the same value, see
        # https://github.com/numpy/numpy/issues/24050
```
Uncommitted comment improvement.

**Review by h-vetinari — COMMENTED** [review]
I'd like to squash-merge this PR (for backportability) soon-ish. It should be uncontroversial, but would be happy to have a second pair of eyes on it!

**Comment by steppi** [comment]
> I'd like to squash-merge this PR (for backportability) soon-ish. It should be uncontroversial, but would be happy to have a second pair of eyes on it!

Sounds good. I should have time to give it a look in the next couple hours.

**Comment by steppi** [comment]
@h-vetinari, I've only learned this recently, but the default behavior for ufuncs in NumPy and SciPy is to return a scalar when a zero dimensional array is passed

```python
import numpy as np
import scipy.special as sc

np.exp(np.array(2))
# 7.38905609893065

sc.zeta(np.array(2))
# 1.6449340668482264
```

this has apparently been the established behavior for a long time, so I don't think we should be deviating from it for the factorial functions in SciPy. Unfortunately, though I was pinged in gh-15600 when this was being discussed I was going through a difficult time and unable to participate. I think @person142's PR, gh-10954, gave a correct bug fix. 

Which functions in `special` are returning a zero dimensional array for zero dimensional input now? I think we should be consistent with the NumPy ufunc standard and return scalars for zero dimensional input for all ufuncs in `special`. Your convention may arguably be better, but consistency is important, and it's not feasible to change the behavior for all other existing ufuncs. I would consider instances where ufuncs in `special` deviate from the expected conventions to be bugs, and we should try to fix them all.

I'm very sorry I wasn't available to carefully review the relevant PRs last year. It would have saved us some trouble. There wasn't really a dedicated maintainer for `special` during that period, but for the foreseeable future, I should have the bandwidth to fill that role.

**Comment by steppi** [comment]
Also see the `PyArrayReturn` function in the [Array Scalars](https://numpy.org/doc/stable/reference/c-api/array.html#array-scalars) section of the NumPy C API. 

```
This function checks to see if arr is a 0-dimensional array and, if so, returns the appropriate array scalar. It should be used whenever 0-dimensional arrays could be returned to Python.
```

**Comment by h-vetinari** [comment]
Could you open a new issue to discuss that? I think this PR should go into 1.11.1, as 1.11 has already been released with that behaviour, and this is now the only odd one out (and the state before that was even more inconsistent, i.e. #10954 didn't get applied to `factorial2` or `factorialk`).

I don't want to open Pandora's box on this topic, but I pretty strongly disagree with mangling 0-dim arrays to scalars (and my impression had been that numpy had gotten more disciplined about this too recently), because it creates a discontinuity in the "array-ness" of the output across the different dimensions, which makes programming for arrays generically really hard.

While I can be convinced to break behaviour again if people insist, a PR is not the place for such a discussion.

**Comment by steppi** [comment]
> Could you open a new issue to discuss that? I think this PR should go into 1.11.1, as 1.11 has already been released with that behaviour, and this is now the only odd one out (and the state before that was even more inconsistent, i.e. #10954 didn't get applied to `factorial2` or `factorialk`).
> 
> I don't want to open Pandora's box on this topic, but I pretty strongly disagree with mangling 0-dim arrays to scalars (and my impression had been that numpy had gotten more disciplined about this too recently), because it creates a discontinuity in the "array-ness" of the output across the different dimensions, which makes programming for arrays generically really hard.
> 
> While I can be convinced to break behaviour again if people insist, a PR is not the place for such a discussion.

Ok, agreed then. I think it’s fine to merge as is then since this is the behavior of the other factorial functions and there was consensus on that behavior. I’ll make another issue and try to get input from some NumPy devs.

**Review by steppi — APPROVED** [review]
OK, I think this successfully implements the behavior agreed upon in gh-15600, and brings this into line with the other factorial functions. 

I want to create another issue to discuss both the behavior chosen here and the existing default behavior for other ufuncs in NumPy in SciPy. I hope to get as much input as possible from NumPy devs who are familiar with the reasons behind the current behavior for most other ufuncs.

I don't really have enough context yet to know which behavior should be preferred. I'd like to treat this like one of Chesterton's fences though and figure out why the current default behavior to return scalars when the input is a 0d array exists. If it turns out that the behavior to return 0d arrays in this case is preferable, then there needs to be a serious conversation about whether the behavior needs to change for all ufuncs in NumPy and SciPy, and just what that would entail.

**Comment by h-vetinari** [comment]
Thanks @steppi, I appreciate it! 🙏 

In a way we're lucky that this corner case came up, so we can solve this properly. For context: #15600 did not intend to break any long-standing conventions, just to make the various factorial functions consistent between each other (aligning the behaviour with `_ufuncs._factorial` was not really on my radar; partly because it's a private function, but mostly because I was focussed on `scipy.special.factorial{,2,k}`)

**Pull request merged**
